Books should not render Object Replacement Character

vitorroriz opened the pull request:
#### f37fd4bfde59a030ecd63064e33ba96752009532
<pre>
Books should not render Object Replacement Character
<a href="https://bugs.webkit.org/show_bug.cgi?id=267265">https://bugs.webkit.org/show_bug.cgi?id=267265</a>
<a href="https://rdar.apple.com/117087591">rdar://117087591</a>

Reviewed by Tim Nguyen.

Some existent EPUB books contain undesired
object replacement characters (U+0xFFFC).
This might happen, for example, when exporting
tools wrongly introduce these characters next to
images, for uknown reasons.

Although the real bug comes from a bug in this kind of
exporting tool, this hurts users, since, even if the root
cause of this problem is fixed, there are several
books already published with these characters.

Therefore, we are creating an exception for the Books App.
If we identify that WebKit is running in Books context,
we will skip rendering the object replacement characters.

This won&apos;t solve the problem for other epub reders accessed
from the browser. However, this behavior is consistent with
other browser engines, that also render the object replacement
character, since CSS Font specification doesn&apos;t tell otherwise.

 * Source/WebCore/platform/graphics/FontCascade.h:
 (WebCore::FontCascade::isInvisibleReplacementObjectCharacter):
 (WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
 (WebCore::FontCascade::treatAsZeroWidthSpaceInComplexScript):

Canonical link: <a href="https://commits.webkit.org/272830@main">https://commits.webkit.org/272830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0167085e94c5762025fafc625dbfd851fa76eeb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12011 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35872 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/30258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14368 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9184 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29393 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33710 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10140 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29693 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8829 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8985 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29670 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37203 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30205 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/30047 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35091 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7052 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32964 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10833 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7705 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9787 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -266,16 +267,27 @@ class FontCascade : public CanMakeWeakPtr<FontCascade>, public CanMakeCheckedPtr
     static CodePath s_codePath;
 
     FontSelector* fontSelector() const;
+    static bool shouldNotRenderReplacementObjectCharacter(char32_t character)
```
Can we name this `isInvisibleReplacementObjectCharacter` ?

vitorroriz committed:
Books should not render Object Replacement Character
https://bugs.webkit.org/show_bug.cgi?id=267265
rdar://117087591

Reviewed by Tim Nguyen.

Some existent EPUB books contain undesired
object replacement characters (U+0xFFFC).
This might happen, for example, when exporting
tools wrongly introduce these characters next to
images, for uknown reasons.

Although the real bug comes from a bug in this kind of
exporting tool, this hurts users, since, even if the root
cause of this problem is fixed, there are several
books already published with these characters.

Therefore, we are creating an exception for the Books App.
If we identify that WebKit is running in Books context,
we will skip rendering the object replacement characters.

This won't solve the problem for other epub reders accessed
from the browser. However, this behavior is consistent with
other browser engines, that also render the object replacement
character, since CSS Font specification doesn't tell otherwise.

 * Source/WebCore/platform/graphics/FontCascade.h:
 (WebCore::FontCascade::isInvisibleReplacementObjectCharacter):
 (WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
 (WebCore::FontCascade::treatAsZeroWidthSpaceInComplexScript):

Canonical link: https://commits.webkit.org/272830@main
